File: frontend/src/test/notificationService.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { notificationService } from '../services/api/notificationService';
import apiClient from '../services/api/apiClient';

// Mock apiClient
vi.mock('../services/api/apiClient', () => ({
  default: {
    get: vi.fn(),
    put: vi.fn(),
  },
}));

describe('notificationService', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('getNotifications', () => {
    it('should fetch all notifications', async () => {
      const mockNotifications = [
        {
          id: '1',
          type: 'OFFER_RECEIVED',
          title: 'New Offer',
          message: 'You received a new offer',
          isRead: false,
          createdAt: '2025-02-04T12:00:00Z',
        },
        {
          id: '2',
          type: 'MESSAGE_RECEIVED',
          title: 'New Message',
          message: 'You have a new message',
          isRead: true,
          createdAt: '2025-02-03T10:00:00Z',
        },
      ];

      (apiClient.get as ReturnType<typeof vi.fn>).mockResolvedValue({ data: mockNotifications });

      const result = await notificationService.getNotifications();

      expect(apiClient.get).toHaveBeenCalledWith('/notifications');
      expect(result).toEqual(mockNotifications);
    });
  });

  describe('getUnreadCount', () => {
    it('should fetch unread notification count', async () => {
      (apiClient.get as ReturnType<typeof vi.fn>).mockResolvedValue({ data: { count: 5 } });

      const result = await notificationService.getUnreadCount();

      expect(apiClient.get).toHaveBeenCalledWith('/notifications/unread-count');
      expect(result).toBe(5);
    });
  });

  describe('markAsRead', () => {
    it('should mark a notification as read', async () => {
      (apiClient.put as ReturnType<typeof vi.fn>).mockResolvedValue({ data: { success: true } });

      const result = await notificationService.markAsRead('notif-123');

      expect(apiClient.put).toHaveBeenCalledWith('/notifications/notif-123/read');
      expect(result).toBe(true);
    });
  });

  describe('markAllAsRead', () => {
    it('should mark all notifications as read', async () => {
      (apiClient.put as ReturnType<typeof vi.fn>).mockResolvedValue({ data: { markedCount: 3 } });

      const result = await notificationService.markAllAsRead();

      expect(apiClient.put).toHaveBeenCalledWith('/notifications/read-all');
      expect(result).toBe(3);
    });
  });
});

File: backend/src/test/java/com/platform/notification/service/NotificationServiceTest.java
package com.platform.notification.service;

import com.platform.notification.model.Notification;
import com.platform.notification.model.NotificationDTO;
import com.platform.notification.model.NotificationType;
import com.platform.notification.repository.NotificationRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.context.ActiveProfiles;

import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;

@SpringBootTest
@ActiveProfiles("test")
class NotificationServiceTest {

    @Autowired
    private NotificationService notificationService;

    @Autowired
    private NotificationRepository notificationRepository;

    @MockBean
    private SimpMessagingTemplate messagingTemplate;

    private UUID testUserId;

    @BeforeEach
    void setUp() {
        notificationRepository.deleteAll();
        testUserId = UUID.randomUUID();
    }

    @Test
    void createNotification_ShouldCreateAndReturnNotification() {
        Notification notification = notificationService.createNotification(
            testUserId,
            NotificationType.OFFER_RECEIVED,
            "New Offer",
            "You received a new investment offer",
            UUID.randomUUID()
        );

        assertNotNull(notification);
        assertNotNull(notification.getId());
        assertEquals(testUserId, notification.getUserId());
        assertEquals(NotificationType.OFFER_RECEIVED, notification.getType());
        assertEquals("New Offer", notification.getTitle());
        assertFalse(notification.isRead());
    }

    @Test
    void createNotification_ShouldSendWebSocketMessage() {
        notificationService.createNotification(
            testUserId,
            NotificationType.MESSAGE_RECEIVED,
            "New Message",
            "You have a new message",
            UUID.randomUUID()
        );

        verify(messagingTemplate).convertAndSendToUser(
            eq(testUserId.toString()),
            eq("/queue/notifications"),
            any(NotificationDTO.class)
        );
    }

    @Test
    void getNotificationsByUserId_ShouldReturnUserNotifications() {
        // Create multiple notifications
        notificationService.createNotification(testUserId, NotificationType.OFFER_RECEIVED, "Offer 1", "Message 1", null);
        notificationService.createNotification(testUserId, NotificationType.OFFER_ACCEPTED, "Offer 2", "Message 2", null);
        notificationService.createNotification(UUID.randomUUID(), NotificationType.SYSTEM, "Other", "Other user", null);

        List<NotificationDTO> notifications = notificationService.getNotificationsByUserId(testUserId);

        assertEquals(2, notifications.size());
    }

    @Test
    void getUnreadCount_ShouldReturnCorrectCount() {
        notificationService.createNotification(testUserId, NotificationType.OFFER_RECEIVED, "Title 1", "Message 1", null);
        notificationService.createNotification(testUserId, NotificationType.OFFER_ACCEPTED, "Title 2", "Message 2", null);
        notificationService.createNotification(testUserId, NotificationType.MESSAGE_RECEIVED, "Title 3", "Message 3", null);

        long count = notificationService.getUnreadCount(testUserId);

        assertEquals(3, count);
    }

    @Test
    void markAsRead_ShouldMarkNotificationAsRead() {
        Notification notification = notificationService.createNotification(
            testUserId,
            NotificationType.OFFER_RECEIVED,
            "Title",
            "Message",
            null
        );

        boolean result = notificationService.markAsRead(notification.getId());

        assertTrue(result);
        assertEquals(0, notificationService.getUnreadCount(testUserId));
    }

    @Test
    void markAllAsRead_ShouldMarkAllUserNotificationsAsRead() {
        notificationService.createNotification(testUserId, NotificationType.OFFER_RECEIVED, "Title 1", "Message 1", null);
        notificationService.createNotification(testUserId, NotificationType.OFFER_ACCEPTED, "Title 2", "Message 2", null);
        notificationService.createNotification(testUserId, NotificationType.MESSAGE_RECEIVED, "Title 3", "Message 3", null);

        assertEquals(3, notificationService.getUnreadCount(testUserId));

        int markedCount = notificationService.markAllAsRead(testUserId);

        assertEquals(3, markedCount);
        assertEquals(0, notificationService.getUnreadCount(testUserId));
    }

    @Test
    void notifyOfferReceived_ShouldCreateCorrectNotification() {
        UUID offerId = UUID.randomUUID();
        notificationService.notifyOfferReceived(testUserId, "John Investor", "My Startup", offerId);

        List<NotificationDTO> notifications = notificationService.getNotificationsByUserId(testUserId);

        assertEquals(1, notifications.size());
        assertEquals(NotificationType.OFFER_RECEIVED, notifications.get(0).getType());
        assertEquals("New Investment Offer", notifications.get(0).getTitle());
        assertTrue(notifications.get(0).getMessage().contains("John Investor"));
        assertTrue(notifications.get(0).getMessage().contains("My Startup"));
    }

    @Test
    void notifyOfferAccepted_ShouldCreateCorrectNotification() {
        UUID offerId = UUID.randomUUID();
        notificationService.notifyOfferAccepted(testUserId, "TechVenture AI", offerId);

        List<NotificationDTO> notifications = notificationService.getNotificationsByUserId(testUserId);

        assertEquals(1, notifications.size());
        assertEquals(NotificationType.OFFER_ACCEPTED, notifications.get(0).getType());
        assertTrue(notifications.get(0).getMessage().contains("accepted"));
    }
}
